### ISSUE TITLE
DialogWindow closes immediately in VST2 on macOS

### ISSUE THREAD

**Issue by beschulz** [julianstorer/JUCE#94]
Please include:

  1) detailed steps on how to reproduce the bug, preferably with already
     existing JUCE code such as the demo plugin or the demo code

Change the GainPlugin createEditor() to:
```c++
AudioProcessorEditor* createEditor() override {
    DialogWindow::LaunchOptions o;
    o.content.setOwned (new TextButton("btn"));
    o.dialogTitle                   = TRANS("Log In");
    o.content->setBounds(0,0,200,200);
    o.dialogBackgroundColour        = Colours::red;
    o.escapeKeyTriggersCloseButton  = false;
    o.useNativeTitleBar             = false;
    o.resizable                     = false;
    o.launchAsync();
    // note: you could also do that in the constructor of the AudioProcessorEditor, the effect is the same
    return new GenericEditor (*this);
}
```

  2) the expected behaviour
    A DialogWindow is shown

  3) the operating system
    macOS 10.11.5

  4) the architecture (32-bit or 64-bit)
    64-bit

  5) a stack trace
```c++
	SuperSecretProject`juce::ModalComponentManager::ModalItem::cancel(this=0x0000000100c1ec20) + 30 at juce_ModalComponentManager.cpp:65
	SuperSecretProject`juce::ModalComponentManager::endModal(this=0x000000010087b220, component=0x0000000100cbe6d0, returnValue=0) + 120 at juce_ModalComponentManager.cpp:142
	SuperSecretProject`juce::Component::exitModalState(this=0x0000000100cbe6d0, returnValue=0) + 96 at juce_Component.cpp:1756
	SuperSecretProject`JuceVSTWrapper::deleteEditor(this=0x000000010103f000, canDeleteLaterIfModal=true) + 258 at juce_VST_Wrapper.cpp:1376
	SuperSecretProject`JuceVSTWrapper::dispatcher(this=0x000000010103f000, opCode=14, index=0, value=0, ptr=0x000000010086e970, opt=0) + 277 at juce_VST_Wrapper.cpp:1428
	SuperSecretProject`AudioEffect::dispatchEffectClass(e=0x000000010103f030, opCode=14, index=0, value=0, ptr=0x000000010086e970, opt=0) + 211 at audioeffect.cpp:54
	Plugin Host`juce::VSTPluginInstance::dispatch(int, int, long long, void*, float) const + 150
	Plugin Host`juce::VSTPluginWindow::openPluginWindow(void*) + 114
	Plugin Host`juce::Component::sendVisibilityChangeMessage() + 69
	Plugin Host`juce::Component::setVisible(bool) + 292
	Plugin Host`juce::VSTPluginWindow::VSTPluginWindow(juce::VSTPluginInstance&) + 263
	Plugin Host`juce::VSTPluginInstance::createEditor() + 50
	Plugin Host`juce::AudioProcessor::createEditorIfNeeded() + 63
	Plugin Host`PluginWindow::getWindowFor(juce::AudioProcessorGraph::Node*, PluginWindow::WindowFormatType) + 95
	Plugin Host`FilterComponent::mouseUp(juce::MouseEvent const&) + 77
	Plugin Host`juce::Component::internalMouseUp(juce::MouseInputSource, juce::Point<float>, juce::Time, juce::ModifierKeys) + 544
	Plugin Host`juce::MouseInputSourceInternal::sendMouseUp(juce::Component&, juce::Point<float>, juce::Time, juce::ModifierKeys) + 96
	Plugin Host`juce::MouseInputSourceInternal::setButtons(juce::Point<float>, juce::Time, juce::ModifierKeys) + 278
	Plugin Host`juce::MouseInputSourceInternal::handleEvent(juce::ComponentPeer&, juce::Point<float>, juce::Time, juce::ModifierKeys, float) + 355
	Plugin Host`juce::MouseInputSource::handleEvent(juce::ComponentPeer&, juce::Point<float>, long long, juce::ModifierKeys, float) + 90
	Plugin Host`juce::ComponentPeer::handleMouseEvent(int, juce::Point<float>, juce::ModifierKeys, float, long long) + 113
	Plugin Host`juce::NSViewComponentPeer::sendMouseEvent(NSEvent*) + 132
	Plugin Host`juce::NSViewComponentPeer::redirectMouseUp(NSEvent*) + 84
	Plugin Host`juce::JuceNSViewClass::asyncMouseUp(objc_object*, objc_selector*, NSEvent*) + 50
	AppKit`-[NSWindow _handleMouseUpEvent:isDelayedEvent:] + 119
	AppKit`-[NSWindow _reallySendEvent:isDelayedEvent:] + 212
	AppKit`-[NSWindow sendEvent:] + 517
	AppKit`-[NSApplication sendEvent:] + 2540
	AppKit`-[NSApplication run] + 796
	Plugin Host`juce::MessageManager::runDispatchLoop() + 81
	Plugin Host`juce::JUCEApplicationBase::main() + 53
	Plugin Host`juce::JUCEApplicationBase::main(int, char const**) + 158
	Plugin Host`start + 52
```
  6) the plugin format (VST2, VST3, AU, AAX, RTAS) - if applicable
does not work with VST2, works fine with VST3 and AU

  7) which DAW you observed the bug in - if applicable
Juce Plugin Host and Ableton Live 9

I'm on 3592d619795a7e00bab0a5541123e423cec1eff6

The reason for that appears to be, that JuceVSTWrapper::deleteEditor(canDeleteLaterIfModal=true) is called *after* the AudioProcessorEditor is created. 

See https://github.com/julianstorer/JUCE/blob/630ab88f8bf11c8389f99a9ffb6daa3a0e891f14/modules/juce_audio_processors/format_types/juce_VSTPluginFormat.cpp#L2114

**Comment by tpoole** [comment]
I would strongly suggest using a child component to display your dialog; popping up modal dialogs from within plug-ins is fraught with danger. For example, in your code above, if you close the host application while the dialog is open then you will have a memory leak.

**Comment by beschulz** [comment]
Thanks for the suggestion.

The code above is just a minimalistic reproduction of the problem, in reality I'm creating the ModalDialog in the constructor of my PluginEditor and that is managing it's livetime. If I do it slightly later, all is good. Also with VST3 and AU, all is dandy.

AFAIK the modal component manager takes care of closing all remaining modal dialogs at exit. Otherwise PopupMenus et.al. would have the same problem, as they also are modal components (see void ComboBox::showPopup(), PopupMenu::showMenuAsync(), ).

I think it's just an issue with the order stuff is done when a new VST2-Plugin Editor is created.

**Comment by tpoole** [comment]
Popups are vulnerable to the same problem; the JUCE Demo was only fixed a few days ago (6547452f and e2dc4609).

I'll see if we can improve the modal state handling when a VST editor is created, though if I recall correctly it's occasionally necessary to create an editor silently in the background to please some hosts.

**Comment by beschulz** [comment]
I re-designed my UI, so that I no longer have this issue. thx.
